libpromises/evalfunction.c

```diff
@@ -7085,6 +7086,64 @@ static FnCallResult FnCallProcessExists(ARG_UNUSED EvalContext *ctx, ARG_UNUSED
 
 /*********************************************************************/
 
+static FnCallResult FnCallNetworkConnections(EvalContext *ctx, ARG_UNUSED const Policy *policy, const FnCall *fp, ARG_UNUSED const Rlist *finalargs)
+{
+    const char *procdir = getenv("CFENGINE_TEST_OVERRIDE_PROCDIR");
+    if (NULL == procdir)
+    {
+        procdir = "";
+    }
+    else
+    {
+        Log(LOG_LEVEL_VERBOSE, "%s: Overriding /proc location to be %s", fp->name, procdir);
+    }
+
+    JsonElement *json = JsonObjectCreate(5);
+    const char* ports_regex = "^\\s*\\d+:\\s+(?<raw_local>[0-9A-F:]+)\\s+(?<raw_remote>[0-9A-F:]+)\\s+(?<raw_state>[0-9]+)";
+
+    JsonElement *data = NULL;
+    Buffer *pbuf = BufferNew();
+
+    BufferPrintf(pbuf, "%s/proc/net/tcp", procdir);
+    data = GetProcFileInfo(ctx, BufferData(pbuf), NULL, NULL, (ProcPostProcessFn) &NetworkingPortsPostProcessInfo, ports_regex);
+    if (NULL != data)
+    {
+        JsonObjectAppendElement(json, "tcp", data);
+    }
+
+    BufferPrintf(pbuf, "%s/proc/net/tcp6", procdir);
+    data = GetProcFileInfo(ctx, BufferData(pbuf), NULL, NULL, (ProcPostProcessFn) &NetworkingPortsPostProcessInfo, ports_regex);
+    if (NULL != data)
+    {
+        JsonObjectAppendElement(json, "tcp6", data);
+    }
+
+    BufferPrintf(pbuf, "%s/proc/net/udp", procdir);
+    data = GetProcFileInfo(ctx, BufferData(pbuf), NULL, NULL, (ProcPostProcessFn) &NetworkingPortsPostProcessInfo, ports_regex);
+    if (NULL != data)
+    {
+        JsonObjectAppendElement(json, "udp", data);
+    }
+
+    BufferPrintf(pbuf, "%s/proc/net/udp6", procdir);
+    data = GetProcFileInfo(ctx, BufferData(pbuf), NULL, NULL, (ProcPostProcessFn) &NetworkingPortsPostProcessInfo, ports_regex);
+    if (NULL != data)
+    {
+        JsonObjectAppendElement(json, "udp6", data);
+    }
```

Comment:
This stuff is still fairly Linux specific. Maybe this parsing stuff should be moved into `unix_iface.c` and this function will just call it?